src/main/java/org/opensearch/security/authtoken/jwt/JwtVendor.java

```diff
@@ -0,0 +1,151 @@
+import org.apache.cxf.rs.security.jose.jwt.JoseJwtProducer;
+import org.apache.cxf.rs.security.jose.jwt.JwtClaims;
+import org.apache.cxf.rs.security.jose.jwt.JwtToken;
+
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.cxf.rs.security.jose.jwt.JwtUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.OpenSearchSecurityException;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.transport.TransportAddress;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.security.support.ConfigConstants;
+import org.opensearch.security.user.User;
+import org.opensearch.security.securityconf.ConfigModel;
+import org.opensearch.threadpool.ThreadPool;
+
+
+public class JwtVendor {
+    private static final Logger logger = LogManager.getLogger(JwtVendor.class);
+
+    private static JsonMapObjectReaderWriter jsonMapReaderWriter = new JsonMapObjectReaderWriter();
+
+    private JsonWebKey signingKey;
+    private JoseJwtProducer jwtProducer;
+    private ConfigModel configModel;
+    private ThreadContext threadContext;
+
+    public JwtVendor(Settings settings) {
+        JoseJwtProducer jwtProducer = new JoseJwtProducer();
+        try {
+            this.signingKey = createJwkFromSettings(settings);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+        this.jwtProducer = jwtProducer;
+    }
+
+    static JsonWebKey createJwkFromSettings(Settings settings) throws Exception {
+        String signingKey = settings.get("signing_key");
+
+        if (!Strings.isNullOrEmpty(signingKey)) {
+
+            JsonWebKey jwk = new JsonWebKey();
+
+            jwk.setKeyType(KeyType.OCTET);
+            jwk.setAlgorithm("HS512");
+            jwk.setPublicKeyUse(PublicKeyUse.SIGN);
+            jwk.setProperty("k", signingKey);
+
+            return jwk;
+        } else {
+            Settings jwkSettings = settings.getAsSettings("jwt").getAsSettings("key");
+
+            if (jwkSettings.isEmpty()) {
+                throw new Exception(
+                        "Settings for key is missing. Please specify at least the option signing_key with a shared secret.");
+            }
+
+            JsonWebKey jwk = new JsonWebKey();
+
+            for (String key : jwkSettings.keySet()) {
+                jwk.setProperty(key, jwkSettings.get(key));
+            }
+
+            return jwk;
+        }
+    }
+
+    //TODO:Getting roles from User
+    public Map<String, String> prepareClaimsForUser(User user, ThreadPool threadPool) {
+        Map<String, String> claims = new HashMap<>();
+        this.threadContext = threadPool.getThreadContext();
+        final TransportAddress caller = threadContext.getTransient(ConfigConstants.OPENDISTRO_SECURITY_REMOTE_ADDRESS);
+        Set<String> mappedRoles = mapRoles(user, caller);
```

Comment:
What is the caller being used for? Seems like we can extract the roles just from the user no?

Reply:
After check with @cwperks, I think the caller is for one of ways to map roles to host. It contains the hostname and ip address. But still, this entire role method should not be locate in this JwtVendor class in the future. Since we haven't map having a consensus on the solution of payload, I will just leave it here for now.

Reply:
@scrawfor99 caller is used when calculating the roles to use for privilege evaluation. See details on role resolution here: https://github.com/opensearch-project/security/issues/2573#issuecomment-1478172731

When serializing the User for common-utils it performs the roles resolution so that plugins can later on do roles injection to evaluate a user's permissions asynchronously.

When thinking of backwards compatibility of replacing a plugin as an extension, in order for it to be a drop in replacement we would need to perform the roles resolution before the request is sent to the extension and pass those roles somehow (could be through claims in the JWT as presented here)

This function serializes the Security plugin's User object onto a special header in the ThreadContext that is readable by common-utils: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java#L198-L210

The `mappedRoles` param would be the same as Ryan has here.